Services/APIVersioning.md

```diff
@@ -0,0 +1,144 @@
+# TL;DR
+
+* REST APIs are not explicitly versioned.
+* Clients are responsible for ensuring forwards compatibility with future versions of services.
+* Services are responsible for maintaining backwards compatibility of published APIs.
+* Follow Postel’s Law. Be conservative in what you send, and liberal in what you accept from others.
+
+# Table of Contents
+
+* [Goals](#goals)
+* [API Versioning](#api-versioning)
+* [Responsibilities of Third-Party Clients](#responsibilities-of-third-party-clients)
+* [Responsibilities of Wikia-Authored Clients](#responsibilities-of-wikia-authored-clients)
+* [Responsibilities of API Developers](#responsibiltiies-of-api-developers)
+* [A Note on HATEOAS](#a-note-on-hateoas)
+
+# Goals
+
+We want to ensure the following:
+
+* Services can be updated independently, at any time
+* Multiple versions of the same service can run side-by-side (during upgrades, downgrades or during A/B testing)
+* API Clients are easy to develop and maintain
+
+We want to avoid the following:
+
+* Production outages due to API changes
+* Dependencies between upgrades
+* Breaking external clients, both official and unofficial
+
+# API Versioning
+
+APIs should not put any version information in their URI. The only exception to this rule is if the API is a transparent proxy for a third-party API that already does not follow this rule.
+
+Placing the version number in the API encourages backwards-incompatible changes through the draconic measure of copying the entire API, _including all the resources that weren't changed_ to a new location. This in turn creates the additional development, testing and support burden of having to maintain independently a stack of redundant APIs. Eventually the decision is made to limit maintenance costs by sunsetting old versions, and clients that would never have broken because the APIs they relied on were not the ones that caused the version bump, cease to work for no good reason.
+
+This complexity only increases when an API is composed from a number of cooperating services, and we would have to choose between clients having to track independent versions both of multiple sub-APIs (and the versions of dependencies between those APIs), or us having to revise the global version of _every_ API whenever an individual service changes.
+
+(All of this is also applies to performing version negotiation in HTTP headers.)
+
+A much better approach is to do your best to ensure APIs are backwards compatible, and if it becomes necessary necessary, create new resources at new locations only for those things that actually changed.
```

Comment:
@harnash 

(1) Replacing significant resources entirely should be a rare occurrence, and should be something we particularly avoid if the resource is linked from a lot of places. Client services that republish links to these resources have the option of either _also_ duplicating the resources, or providing links to both old and new in existing resources. Old resources can also be modified safely to have "link rel="alternate""-style properties that direct clients to their newer representation, so a client never gets "lost" because it's been linked to an arbitrary version of a resource.

If you think about it, the problem is just as bad, or worse in the "version everything" approach: you're solving "this finite number of resources might need to be duplicated" by replacing it with "every resource needs to be duplicated". Which means that instead of dealing with the cascading effects of a single resource changing, you're dealing with the effects of _every_ resource in that service changing!

Say you have service A that bumps its version number because it has changed the meaning of resource A1, and service B that links to resource A2. In an un-versioned world, B would have to do nothing. In a versioned world, now B has to worry about which version of A2 to link to, potentially forcing B to bump its version too… and so on.

By giving us the ability to update resources, and biasing us against breaking them, we can deliberately control the blast-radius of significant changes in a way that Version All The Things does not permit.

(2) If the data model changes in a way that is not backwards compatible, we have a problem _however_ we choose to version our APIs, as we would still have the problem of supporting the old data model in the previous version of the API.

In a versioned-API world, this would leave us with the choice of removing the old version of the API because it represented a data model we can no longer express, or continuing to publish the API in a backwards-incompatible state. The former is actually a worse state of affairs for clients than not versioning at all, the latter is really the same as not versioning the API at all.